include/nbl/video/surface/ISurface.h

```diff
@@ -2,54 +2,80 @@
 
 #include "nbl/core/IReferenceCounted.h"
+
 #include "nbl/asset/format/EColorSpace.h"
 #include "nbl/asset/format/EFormat.h"
 
+#include "nbl/video/IAPIConnection.h"
+
+#define VK_NO_PROTOTYPES
+#include "vulkan/vulkan.h"
+
 namespace nbl::video
 {
 
 class IPhysicalDevice;
 
 class ISurface : public core::IReferenceCounted
 {
-protected:
-    virtual ~ISurface() = default;
-
-public:
-    struct SColorSpace
-    {
-        asset::E_COLOR_PRIMARIES primary;
-        asset::ELECTRO_OPTICAL_TRANSFER_FUNCTION eotf;
-    };
-    struct SFormat
-    {
-        asset::E_FORMAT format;
-        SColorSpace colorSpace;
-    };
-    enum E_PRESENT_MODE
-    {
-        EPM_IMMEDIATE = 0,
-        EPM_MAILBOX = 1,
-        EPM_FIFO = 2,
-        EPM_FIFO_RELAXED = 3
-    };
-
-    enum E_SURFACE_TRANSFORM_FLAGS : uint32_t
-    {
-        EST_IDENTITY_BIT = 0x00000001,
-        EST_ROTATE_90_BIT = 0x00000002,
-        EST_ROTATE_180_BIT = 0x00000004,
-        EST_ROTATE_270_BIT = 0x00000008,
-        EST_HORIZONTAL_MIRROR_BIT = 0x00000010,
-        EST_HORIZONTAL_MIRROR_ROTATE_90_BIT = 0x00000020,
-        EST_HORIZONTAL_MIRROR_ROTATE_180_BIT = 0x00000040,
-        EST_HORIZONTAL_MIRROR_ROTATE_270_BIT = 0x00000080,
-        EST_INHERIT_BIT = 0x00000100,
-        EST_FLAG_BITS_MAX_ENUM = 0x7FFFFFFF
-    };
-
-    // vkGetPhysicalDeviceSurfaceSupportKHR on vulkan
-    virtual bool isSupported(const IPhysicalDevice* dev, uint32_t _queueFamIx) const = 0;
+    protected:
+        ISurface(core::smart_refctd_ptr<IAPIConnection>&& api) : m_api(std::move(api)) {}
+        ~ISurface() = default;
+
+        core::smart_refctd_ptr<IAPIConnection> m_api;
+
+    public:
+        struct SColorSpace
+        {
+            asset::E_COLOR_PRIMARIES primary;
+            asset::ELECTRO_OPTICAL_TRANSFER_FUNCTION eotf;
+        };
+        struct SFormat
+        {
+            asset::E_FORMAT format;
+            SColorSpace colorSpace;
+        };
+        enum E_PRESENT_MODE
+        {
+            EPM_IMMEDIATE = 0,
+            EPM_MAILBOX = 1,
+            EPM_FIFO = 2,
+            EPM_FIFO_RELAXED = 3,
+            EPM_UNKNOWN = 4
+        };
+
+        struct SCapabilities
+        {
+            uint32_t minImageCount;
+            uint32_t maxImageCount;
+            VkExtent2D currentExtent;
+            VkExtent2D minImageExtent;
+            VkExtent2D maxImageExtent;
+            uint32_t maxImageArrayLayers;
+            // Todo(achal)
+            // VkSurfaceTransformFlagsKHR       supportedTransforms;
+            // VkSurfaceTransformFlagBitsKHR    currentTransform;
+            // VkCompositeAlphaFlagsKHR         supportedCompositeAlpha;
+            asset::IImage::E_USAGE_FLAGS supportedUsageFlags;
+        };
+
+        inline E_API_TYPE getAPIType() const { return m_api->getAPIType(); }
+
+        // vkGetPhysicalDeviceSurfaceSupportKHR on vulkan
+        virtual bool isSupported(const IPhysicalDevice* dev, uint32_t _queueFamIx) const = 0;
+
```

Comment:
we make all the surface vs physical device methods be methods of the swapchain (makes it easy to have spoof surfaces)